### ISSUE TITLE
Memory leak for restore and input pipe -- FIXED

### ISSUE THREAD

**Issue by 0bi-w6n-K3nobi** [allinurl/goaccess#1929]
I tried hard to find this bug for 2 weeks. So simple but so difficult to find where.
I used several tools for that, like ` valgrind`and `CLang` with `AddressSanitizer`.
I managed to reduce my test to an `LOG` with 4 lines and found the leak.
In 4th record than occurs the error.

How to reproduce the problem:

`mkdir db`
`echo -n '' | goaccess --persist --db-path=./db --output=./a.html -`
`cat logfile | goaccess`**`--restore`**`--persist --db-path=./db --output=./a.html`**`-`**

But, In another way, **no error** occurs:

`mkdir db`
`echo -n '' | goaccess --persist --db-path=./db --output=./a.html -`
`goaccess`**`--restore`**`--persist --db-path=./db --output=./a.html`**`logfile`**
 
or 

`mkdir db`
`echo -n '' | goaccess --persist --db-path=./db --output=./a.html -`
`cat logfile | goaccess --persist --db-path=./db --output=./a.html`**`-`**
 
So I found the code below,  at `parse.c` around `2530`:

` ` `/* it's a pipe, then use the last parsed timestamp */`
` ` `ts = mktime (&logitem->dt);`
` ` `if (conf.restore && !glog->inode && last > 0 && last >= ts)`
` ` ` ` ` ` **`return 0;`**

Oops! Maybe should must be:

` ` `/* it's a pipe, then use the last parsed timestamp */`
` ` `ts = mktime (&logitem->dt);`
` ` `if (conf.restore && !glog->inode && last > 0 && last >= ts)`**`{`**
` ` ` ` ` ` **`ret = 0;`**
` ` ` ` ` ` **`goto cleanup;`**
` ` **`}`**

I thought and rethought several times ... so I removed the code, and it became:

` ` `ts = mktime (&logitem->dt);`

I believe that there should be no differences between `PIPE` and normal `LOG` file.
It is very common for me to apply a filter with `awk` and use the `PIPE`, even for old `LOGs` [ offline ].
I think that you should have used this conditional as an **optimization**, for **real-time** processing.

**PS**: In my `LOG` of 1.8 million records I had a leak of more than **2 Gigabytes** of memory.

All Right. See you soon.

**Comment by allinurl** [comment]
Thanks for pointing that out. I can certainly apply this fix. Though, as far as removing that check, I'm not sure if we can, relying only on the timestamp wouldn't be enough.  I'm working on a fix for #1815 that relates to this.

**Comment by allinurl** [comment]
This should be fixed. Thanks again for reporting it!

**Comment by 0bi-w6n-K3nobi** [comment]
@allinurl, good morning.
Before everything, it was a great job for patch 2e14dc3. I believe that you worked a lot for that.

But, although it is appropriate for most cases and users, this will not be enough for me.
For you to understand me better, I will quote again a practical example.
And again using the `LOG` of issue #1927.
Please, be patient and read until the **solution**.

Discribing more detaily, this `LOG` is a concat of serveral small `LOGs`, from a NGInX server in front a WordPress backend.
So, the final `LOG` contains mixing of records with diferents `VHOSTS`, and certainly **not ordered in time** among itselves.
That it the `BIG` problem.

So running, it will show:

`goaccess -o ./a.html LOGFILE`  or
`cat LOGFILE | goaccess -o ./a.html -`

![image](https://user-images.githubusercontent.com/71027865/95690931-9b921c80-0bf1-11eb-83bc-80afeb66f625.png)

Soon after, I break the `LOGFILE` in several parts, each one for each different `VHOST`.
So, for each log we have a common log with records ordered in time.
In fact, there are exactly 34 `VHOSTS`. And again, I show same graphic above, for:

`goaccess -o ./a.html a-site-log b c d ... `  or
`cat a-site-log b c d ... | goaccess -o ./a.html -`

So simple until here. Ok, going to save and restore problem.
If I running follow command, We will see same graphic above:

`mkdir db`
`echo -n '' | goaccess --persist -o ./b.html -`
`goaccess --restore --persist -o ./b.html`**`a`**
`goaccess --restore --persist -o ./b.html`**`b`**
`goaccess --restore --persist -o ./b.html`**`c`**
`...`

I.E. no problem until here. But, if i running:

`mkdir db`
`echo -n '' | goaccess --persist -o ./b.html -`
`cat`**`a`**`| goaccess --restore --persist -o ./b.html`**`-`**
`cat`**`b`**`| goaccess --restore --persist -o ./b.html`**`-`**
`cat`**`c`**`| goaccess --restore --persist -o ./b.html`**`-`**
`...`
or in same way, same result:

`mkdir db`
`echo -n '' | goaccess --persist -o ./b.html -`
`cat`**`a b c d ...`**`| goaccess --restore --persist -o ./b.html`**`-`**

![image](https://user-images.githubusercontent.com/71027865/95692955-e87cef80-0bff-11eb-9c08-fc1563533693.png)

Other result, can see follow:

`mkdir db`
`cat $(\ls -1S  a b c d .. | head -1) | goaccess --persist -o ./b.html -`
`cat $(\ls -1S  a b c d .. | tail -n +2)  | goaccess --restore --persist -o ./b.html`**`-`**

![image](https://user-images.githubusercontent.com/71027865/95757086-0b011e00-0c7d-11eb-86bf-e1e9350d40bb.png)

And finally for original `LOGFILE`, follow you see:

`mkdir db`
`echo -n '' | goaccess --persist -o ./b.html -`
`cat LOGFILE | goaccess --restore --persist -o ./b.html`**`-`**

![image](https://user-images.githubusercontent.com/71027865/95694375-2c73f280-0c08-11eb-8f12-22e88601b69e.png)

**the problem** [ context ]:
If we have a set of `LOGs` that need account in unique report, it is not possible through that input `PIPE` with using `--restore` option.
Input `PIPE` is good and nescessary option for compressed `LOGs` or if need apply some filter through  `AWK` or `SED` tools.
In examples above, I'm using single-day, but of course mainly use is for accumulated day-by-day counting.

**solution**:
The creation of `CONF` flag like `ignore_timestamp` for processing of multiple `VHOSTs` `LOGs` with `--restore` option.
Of course, the default value should be `zero`/`false`, therefore turn compatible with current behavior.
In this way, the code should be:

`static int`
`should_restore_from_disk (GLog * glog) {`
` ` ` ` `GLastParse lp = { 0 };`

` ` ` ` `if (!conf.restore)`
` ` ` ` ` ` ` ` `return 0;`

` ` ` ` **`if (conf.ignore_timestamp)`**
` ` ` ` ` ` ` ` **`return 0;`**

**PS**:
Running `GOACCESS` compiled with `CLANG` with `AddressSanitizer` accused some leaks into `HASH`-tables routines.
I will investigated later, and propos some fix for that.

That's all for now folks.
See you soon.

**Comment by allinurl** [comment]
@0bi-w6n-K3nobi Thanks for posting this. You bring a great point, like you said, the timestamp check with entries that are unordered doesn't really help and will certainly either deflate the output. Just so I understand what you are proposing with `conf.ignore_timestamp`, how would you protect/check that repeated records are not counted twice? or would this be a flag that will process all records but with the consent from the user that it won't check dups?

Thanks again!

**Comment by allinurl** [comment]
By the way, feel free to post any findings from the AddressSanitizer :)

**Comment by allinurl** [comment]
@0bi-w6n-K3nobi I've pushed some changes that fixes a few mem leaks when restoring from disk. Feel free to rebase from origin. Thanks

**Comment by 0bi-w6n-K3nobi** [comment]
You understood me correctly. What I am suggesting is that it is possible to circumvent protection.
It will really be difficult to contemplate all the scenarios.
Have you ever thought that the user can erroneously decompress the same LOG, and because it has the new inode and then process twice?
I can also skip the test/protection by creating a FIFO file every time, which will have a new inode every time, and use it as a PIPE.

My God...
As my brother always said: you must think an algorithm that includes 1 single file or infinite files. Being in the middle is always very bad.

**Comment by 0bi-w6n-K3nobi** [comment]
The problem I showed came up because I am processing different LOGs / VHOSTs for a single input which is the PIPE, or in other words a file with inode == 0.
But, as I exemplified, everything is wonderful if I use a file for each LOG and each one with its own inode.

**Comment by allinurl** [comment]
Got it. Yeah, I think it would be fine to add the `--ignore_timestamp` flag. Thanks for proposing that. Please feel free to post a patch or a PR and happy to merge it.

**Comment by 0bi-w6n-K3nobi** [comment]
Well @allinurl  ... The patch is very simple, as I mentioned in lines of code above.
I also feel awkward about suggesting such a patch. I will like a more elegant solution.
There may be a simple solution, but it must be so obvious that we don't see it.

**Comment by 0bi-w6n-K3nobi** [comment]
Ok, I will test your fixes soon as possible.
And I will do the patch if you insist on it.

Thanks for your availability.

**Comment by 0bi-w6n-K3nobi** [comment]
Yeah. I believe almost leaks vanish.
I ran the updated code until the patch ebec767.
I found one simple:

```
parser.c
@@ -1089,6 +1089,8 @@`
       return spec_err (logitem, SPEC_TOKN_NUL, *p, NULL);
     if (is_cache_hit (tkn))
       logitem->cache_status = tkn;
+    else
+      free (tkn);
     break;
     /* remote hostname (IP only) */
   case 'h':

```

But, CLang with AddressSanitizer found 2 others:

```
Direct leak of 88 byte(s) in 1 object(s) allocated from:
    #0 0x4c3127 in calloc (/home/fabiano/rpmbuild/SRPMS/goaccess-1.4-1/goaccess-git/goaccess+0x4c3127)
    #1 0x60474c in xcalloc /home/fabiano/rpmbuild/SRPMS/goaccess-1.4-1/goaccess-git/src/xmalloc.c:69:14
    #2 0x565286 in new_gmetrics /home/fabiano/rpmbuild/SRPMS/goaccess-1.4-1/goaccess-git/src/gstorage.c:50:23
    #3 0x5256d7 in set_root_metrics /home/fabiano/rpmbuild/SRPMS/goaccess-1.4-1/goaccess-git/src/gholder.c:559:13
    #4 0x520e12 in add_root_to_holder /home/fabiano/rpmbuild/SRPMS/goaccess-1.4-1/goaccess-git/src/gholder.c:581:7
    #5 0x51d5c7 in load_holder_data /home/fabiano/rpmbuild/SRPMS/goaccess-1.4-1/goaccess-git/src/gholder.c:640:5
    #6 0x55f7e4 in allocate_holder_by_module /home/fabiano/rpmbuild/SRPMS/goaccess-1.4-1/goaccess-git/src/goaccess.c:310:3
    #7 0x55c5a5 in allocate_holder /home/fabiano/rpmbuild/SRPMS/goaccess-1.4-1/goaccess-git/src/goaccess.c:321:5
    #8 0x55b9fe in main /home/fabiano/rpmbuild/SRPMS/goaccess-1.4-1/goaccess-git/src/goaccess.c:1472:3
    #9 0x7eff1433d1a2 in __libc_start_main (/lib64/libc.so.6+0x271a2)
```
```
Indirect leak of 8 byte(s) in 1 object(s) allocated from:
    #0 0x4c2f6f in malloc (/home/fabiano/rpmbuild/SRPMS/goaccess-1.4-1/goaccess-git/goaccess+0x4c2f6f)
    #1 0x604444 in xmalloc /home/fabiano/rpmbuild/SRPMS/goaccess-1.4-1/goaccess-git/src/xmalloc.c:46:14
    #2 0x604707 in xstrdup /home/fabiano/rpmbuild/SRPMS/goaccess-1.4-1/goaccess-git/src/xmalloc.c:58:9
    #3 0x52bf8e in get_is32 /home/fabiano/rpmbuild/SRPMS/goaccess-1.4-1/goaccess-git/src/gkhash.c:1840:12
    #4 0x52be59 in ht_get_datamap /home/fabiano/rpmbuild/SRPMS/goaccess-1.4-1/goaccess-git/src/gkhash.c:2619:10
    #5 0x522121 in map_data /home/fabiano/rpmbuild/SRPMS/goaccess-1.4-1/goaccess-git/src/gholder.c:420:19
    #6 0x5254e1 in set_root_metrics /home/fabiano/rpmbuild/SRPMS/goaccess-1.4-1/goaccess-git/src/gholder.c:551:7
    #7 0x520e12 in add_root_to_holder /home/fabiano/rpmbuild/SRPMS/goaccess-1.4-1/goaccess-git/src/gholder.c:581:7
    #8 0x51d5c7 in load_holder_data /home/fabiano/rpmbuild/SRPMS/goaccess-1.4-1/goaccess-git/src/gholder.c:640:5
    #9 0x55f7e4 in allocate_holder_by_module /home/fabiano/rpmbuild/SRPMS/goaccess-1.4-1/goaccess-git/src/goaccess.c:310:3
    #10 0x55c5a5 in allocate_holder /home/fabiano/rpmbuild/SRPMS/goaccess-1.4-1/goaccess-git/src/goaccess.c:321:5
    #11 0x55b9fe in main /home/fabiano/rpmbuild/SRPMS/goaccess-1.4-1/goaccess-git/src/goaccess.c:1472:3
    #12 0x7eff1433d1a2 in __libc_start_main (/lib64/libc.so.6+0x271a2)
```

These 2 leaks happen only 3 times in the processing of the 34 LOGs already mentioned.
And this occurs with only a few bytes as can be seen.

I tried to go through the code and couldn't identify the problem.
I know that the variables [sugested like error] will be freed at the end, when the panel structures are released.
So, why does CLang accuse this like error?

Things and Thinks of Language C.

**Comment by allinurl** [comment]
Thanks for pointing this out. Would you be able to spot the offending lines for these two? Or would you be able to run it through valgrind and see what output do you get? Thanks again!

**Comment by allinurl** [comment]
14c8aa92550d8412f62c593555a22896c15ca375 fixes the leak pointed above. Thank you, sir!

**Comment by 0bi-w6n-K3nobi** [comment]
I will try with `valgrind` and soon that possible give you the result.

**Comment by 0bi-w6n-K3nobi** [comment]
Oooh man! It was so hard to find the problem.
I tried `valgrind` but did not was better than CLang. It is so slow, but ok.
I was able to find the exact record and condition to repeat the leak.
After that, then I tried this trick with gdb with backtraces, found [here](https://stackoverflow.com/questions/311840/tool-to-trace-local-function-calls-in-linux/10146585) and [there](https://www.programmersought.com/article/17623303505/).

I filter just for `BT` only with functions into `goaccess.c`, `gholder.c` and `gstore.c`

I did just compare the behavior with a good request and bad request.
So little bit diference, and no more info. So I gave up on this procedure.

I look with more attention, and examed the leak error. And voala:
```
goaccess-git/src/gholder.c
@@ -581,8 +581,11 @@
   if (set_root_metrics (item, h->module, type, &nmetrics) == 1)
     return;

-  if (!(root = ht_get_root (h->module, item.nkey)))
+  if (!(root = ht_get_root (h->module, item.nkey))) {
+    free (nmetrics->data);
+    free (nmetrics);
     return;
+  }

   /* add data as a child node into holder */
   if (KEY_NOT_FOUND == (root_idx = get_item_idx_in_holder (h, root))) {
```
As I mentioned, it happens rarely. But it is a mistake.

See you soon.
